src/Control/Effect/Replay/TH.hs

```diff
@@ -73,15 +73,24 @@ decodeSingle keyVal valVal con = do
       constructorTupleP :: PatQ
       constructorTupleP = tupP (constructorNameP : constructorArgsP)
 
+  let argBindingsE :: [StmtQ]
+      argBindingsE =
+        if null constructorArgsP
+          then -- No-arg constructors are meaningless to match
+            []
+          else -- (("MyGadt.Bar" :: String), a, b) <- fromRecordedValue keyVal
+            [bindS constructorTupleP [e|fromRecordedValue $(varE keyVal)|]]
+
   -- do
   doE
-    -- (("MyGadt.Bar" :: String), a, b) <- fromRecordedValue keyVal
-    [ bindS constructorTupleP [e|fromRecordedValue $(varE keyVal)|]
-    , -- decodedResultValue <- fromRecordedValue valVal
-      bindS (sigP (varP resultNm) (getGadtTy con)) [e|fromRecordedValue $(varE valVal)|]
-    , -- pure (EffectResult (Bar a b) decodedResultValue)
-      noBindS [e|pure (EffectResult $(conAppliedE) $(varE resultNm))|]
-    ]
+    ( argBindingsE
+        <> [
+             -- decodedResultValue <- fromRecordedValue valVal
+             bindS (sigP (varP resultNm) (getGadtTy con)) [e|fromRecordedValue $(varE valVal)|]
+           , -- pure (EffectResult (Bar a b) decodedResultValue)
+             noBindS [e|pure (EffectResult $(conAppliedE) $(varE resultNm))|]
+           ]
+    )
```

Comment:
Is this change related?  I'd prefer to leave it out if not, TH code is really hard to review, I don't want to block the new effect constructor for this.

Reply:
Oh, I forgot to explain that.  The TH we have doesn't seem to support no-arg constructors.  It ended up with missing the ReplayableData instance on `Unit(String)`.  I'm not sure what that means though, and it didn't seem necessary to match anyway so I did this instead.

I'll take another look tomorrow.

Reply:
I removed the `deriveReplayable` and reverted this, as you suggested.  I don't think it was correct anyway.